### PR TITLE
Zoom layout shift: second alternate fix.

### PR DESCRIPTION
**Pull request by jasmussen**
## What?

Alternative to #65915, and #65967. Removes flex from the zoomed out body element, replaces with a transparent border to prevent margin collapsing from causing a layout shift:

test2

## Testing Instructions

Go to the post editor, zoom out. There should be no layout shift.

### CODE REVIEW

**Review by andrewserong — APPROVED** [review]
Thank you for the persistence here @jasmussen! To me this is the right fix to get in for WP 6.7: it fixes the layout shift and addresses the width issues caused by the presence of the `flex` rule, as @stokesman has highlighted in a couple of places (e.g. https://github.com/WordPress/gutenberg/pull/65627#issuecomment-2406337877)

This PR does effectively revert the change in https://github.com/WordPress/gutenberg/pull/59512 that expanded the main content area while in zoom out mode. I think this is a good trade-off for 6.7 IMO. Here's how it looks in this PR (note the main `Type / to choose a block` area does not expand on zoom out):

https://github.com/user-attachments/assets/65c0fe8c-5fa4-41c9-81e2-ba8d98c2bbee

I'll just add a few other folks as reviewers for a second opinion here in case my assumptions here are at all off, as see there's some related discussion in #65627. Overall, though, I like the idea of this sort of simple CSS change 👍

Adding a tentative approval as I think this would be really good to get in.

File: packages/block-editor/src/components/iframe/content.scss
Comment (kevin940726) [inline packages/block-editor/src/components/iframe/content.scss:3]: A comment here would help the readability here I believe 👍

**Review by kevin940726 — APPROVED** [review]

**Comment by kevin940726** [comment]
I added the backport label since it seems to target the 6.7 release.

**Comment by jasmussen** [comment]
I'm going to go ahead and merge this, because we need this in for zoom to make sense. Hopefully it can be upgraded and improved with additional patches later.

**Comment by madhusudhand** [comment]
This fix is having a side effect on patterns preview.

pattern-rendering

**Comment by t-hamano** [comment]
@madhusudhand Is this problem occurring on the trunk? I've tried several patterns but I can't reproduce it.

However, I've seen this kind of strange behavior a few times in the past.

File: packages/block-editor/src/components/iframe/content.scss
Comment (ellatrix) [inline packages/block-editor/src/components/iframe/content.scss:36]: Why was this removed? I believe this will break filling the page to fit the canvas height, cc @richtabor

Comment (jasmussen) [reply packages/block-editor/src/components/iframe/content.scss:36]: The context is in #65915. Either these rules persist on body, _both zoomed in and zoomed out_, or they are removed. Having them just in one case, causes the layout shift shown in 65915.

Comment (ellatrix) [reply packages/block-editor/src/components/iframe/content.scss:36]: I can reproduce it from the commit right before this, but _I think_ this didn't happen with zoom-out mode with TT4 just a few months ago. Maybe something changed that caused this glitchy behaviour.

zoom-out-glitch

The problem is that now the footer will no longer be pushed to the bottom, so we'll have to revert these lines in particular.

|Before|After|
|-|-|
|Screenshot 2024-10-16 at 17 24 47|Screenshot 2024-10-16 at 17 22 38|

Comment (jasmussen) [reply packages/block-editor/src/components/iframe/content.scss:36]: It's fine in the site editor, it's the _post_ editor that's the problem due to how the Title block has margins that collapse.

Comment (stokesman) [reply packages/block-editor/src/components/iframe/content.scss:36]: > The problem is that now the footer will no longer be pushed to the bottom

I think it was an understood tradeoff. The flex display on the body was also the cause of #63519 so bringing it back brings that back as well.

Comment (ellatrix) [reply packages/block-editor/src/components/iframe/content.scss:36]: This is not a tradeoff 😃 this is removing a feature that I added half a year ago

Comment (draganescu) [reply packages/block-editor/src/components/iframe/content.scss:36]: @ellatrix do we need the rule to apply at all times? Methinks that maybe when there is no template, there is no footer to be pushed down. So then apply flex only when there is a template?

Comment (ellatrix) [reply packages/block-editor/src/components/iframe/content.scss:36]: Yes if the problem is only in the post editor, we should only remove it in the post editor

Comment (stokesman) [reply packages/block-editor/src/components/iframe/content.scss:36]: I didn’t mean to disparage anything by calling it a tradeoff, but fixing a couple things while knowingly breaking another seems to fit the definition to me. Though, yes, perhaps it doesn’t have to break.

Though I find it odd that for that feature we justify layout shift, since that’s what pushing the footer is. It would seem better if there were a way to orchestrate it after the zoom out transition and include some helpful indicator like expanding one of the insertion points.

Comment (jasmussen) [reply packages/block-editor/src/components/iframe/content.scss:36]: Clarifying further, theoretically it could also be the site editor. It's any editor where the first block uses a _top margin_. Any such margin will collapse when switching between flex and not block. That also means in the post editor, if you turn "Show template" on or off, you can switch between maybe having top margins or not:

abstracted

I'd be happy to restore the rule so it exists both zoomed and unzoomed. But the layout shift has to be fixed somehow.

Comment (stokesman) [reply packages/block-editor/src/components/iframe/content.scss:36]: > I'd be happy to restore the rule so it exists both zoomed and unzoomed

Maybe I am missing something but to add a such a rule on the `body` (especially unzoomed) seems contrary to the isolation of UI and theme styling that the iframe is supposed to provide. Do we mandate that block based themes do not control `body` styling? If so, I guess that’s what I'm missing. Otherwise, it seems like a non-starter—i.e. we should not add the rule (at least while unzoomed, but I have my doubts it’s proper even when zoomed).

**Comment by t-hamano** [comment]
In resizable editors, the top border appears to cause an unintended vertical scrollbar. Please see #66297.

**Comment by andrewserong** [comment]
Thanks for flagging @t-hamano, I've opened a tentative fix over in https://github.com/WordPress/gutenberg/pull/66301 — it's a bit of a hack (add `1px` to the resizeable editor to account for sub-pixel height). Happy to close it out if there's a better approach.

File: packages/block-editor/src/components/iframe/content.scss
Comment (ramonjd) [reply packages/block-editor/src/components/iframe/content.scss:3]: Wondering if this should be `outline` rather than border, unless the inset style is preferred?

| Border  (trunk) | Outline |
|--------|--------|
| Screenshot 2024-10-23 at 10 49 24 am | Screenshot 2024-10-23 at 10 49 18 am |

Comment (ramonjd) [reply packages/block-editor/src/components/iframe/content.scss:3]: Actually, ignore me please - refreshing and updating trunk helps 😄

Comment (ramonjd) [reply packages/block-editor/src/components/iframe/content.scss:3]: I'd still use outline - the border is causing some minor layout discrepancies:


https://github.com/user-attachments/assets/f5b2dd18-4ca8-40fc-b6c2-1929e5ebbe79

Comment (stokesman) [reply packages/block-editor/src/components/iframe/content.scss:3]: I don’t think outline can work. The border is to cause margins of children to be contained something outline won’t do. That is this issue will return:

https://github.com/user-attachments/assets/27fff5a9-92fd-40df-bd8d-1b0a92cca820

But it may be we need an alternative to using the border as well. Good spotting this Ramon.

Comment (ramonjd) [reply packages/block-editor/src/components/iframe/content.scss:3]: > The border is to cause margins of children to be contained something outline won’t do.

Ah gotcha, thanks for the explainer.

I was testing in the site editor where the effect wasn't so pronounced. 

I wonder if there's another property to prevent margin collapsing, e.g, `display: flow-root;`

Comment (stokesman) [reply packages/block-editor/src/components/iframe/content.scss:3]: `display: flow-root` is looking good from my testing.

Comment (ciampo) [reply packages/block-editor/src/components/iframe/content.scss:3]: `box-shadow` is also an alternative to display a visual border without causing layout shifts

Comment (ramonjd) [reply packages/block-editor/src/components/iframe/content.scss:3]: I think the goal was to prevent margin collapsing rather than displaying a border. The border in this PR is transparent.

See related: 

- https://github.com/WordPress/gutenberg/pull/66390

File: packages/block-editor/src/components/iframe/content.scss
Comment (stokesman) [reply packages/block-editor/src/components/iframe/content.scss:36]: > The problem is that now the footer will no longer be pushed to the bottom

I’ve opened up #66388 to restore the feature in a different way that should be less liable to have other impacts. I already requested a review from some of the folks here but I'm mentioning it in case anyone else cares to have a look.

**Pull request merged**
